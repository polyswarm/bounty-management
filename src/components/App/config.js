const config = {
  url: process.env.REACT_APP_HOST,
};
export default config;
