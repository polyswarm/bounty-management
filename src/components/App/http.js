import BigNumber from 'bignumber.js';
class HttpApp {
  constructor(url, ws) {
    this.url = url;
    this.ws = ws;
  }

  getUnlockedWallet() {
    return fetch(this.url+'/accounts/active')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => !json.locked)
      .catch(() => false);
  }

  getWallets() {
    return fetch(this.url+'/accounts')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .catch(() => []);
  }

  getBounty(bounty) {
    return fetch(this.url+'/bounties/'+bounty.guid)
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw Error('Cannot get bounties.');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(bounty => this.getAssertionsForBounty(bounty))
      .then(bountyAssertions => this.getArtifactsForBounty(bountyAssertions))
      .then(bounty => {
        const assertions = bounty.assertions;
        let files = bounty.artifacts;
        assertions.forEach((assertion) => {
          assertion.verdicts.forEach((verdict, index) => {
            const file = files[index];
            if (!verdict) {
              file.good++;
            }
            file.total++;
            file.assertions.push({
              author: assertion.author,
              bid: assertion.bid,
              verdict: verdict,
              metadata: assertion.metadata
            });
            files[index] = file;
          });
        });
        bounty.artifacts = files;
        return bounty;
      })
      .catch(() => null);
  }

  getArtifactsForBounty(bounty) {
    return fetch(this.url+'/artifacts/'+bounty.uri)
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(files => {
        return files.map((file) => {
          return file.name;
        });
      })
      .then(filesnames => {
        return filesnames.map((name) => {
          const trimmed = name.trim();
          return {name: trimmed, good: 0, total: 0, assertions: []};
        });
      })
      .then(files => {
        bounty.artifacts = files;
        return bounty;
      });
  }

  getAssertionsForBounty(bounty) {
    return fetch(this.url+'/bounties/'+bounty.guid+'/assertions')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(assertions => {
        return assertions.map((assertion) => {
          const bid = new BigNumber(assertion.bid).dividedBy(new BigNumber('1000000000000000000')).toNumber();
          return {
            author: assertion.author,
            bid: bid,
            verdicts: assertion.verdicts,
            metadata: assertion.metadata,
          };
        });
      })
      .then(filtered => {
        bounty.assertions = filtered;
        return bounty;
      });
  }

  listenForAssertions(assertionAddedCallback) {
    // attach to websocket
    // anytime we get an assertion, check if it matches a guid
    // if it does, add it to the assertions for that object
    const ws = this.ws;
    const websocket = new WebSocket(ws);

    websocket.onmessage = (event) => {
      const message = JSON.parse(event.data);

      if (message.event === 'assertion') {
        const body = message.data;
        const bid = new BigNumber(body.bid).dividedBy(new BigNumber('1000000000000000000')).toNumber();
        const assertion = {
          guid: body.bounty_guid,
          bid: bid,
          verdicts: body.verdicts,
          metadata: body.metadata,
          author: body.author,
        };
        assertionAddedCallback(assertion);
      }
    };
  }
}
export default HttpApp;
