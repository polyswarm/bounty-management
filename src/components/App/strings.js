const strings = {
  create: 'Create a Bounty',
  bountyError: 'Error: Failed to Post Bounty',
  requestAllData: 'Refreshing bounties',
  requestWallets: 'Updating Wallets',
  requestGetBounty: 'Getting Bounty'
};

export default strings;
