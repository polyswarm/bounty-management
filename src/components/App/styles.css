html,
body,
#root {
  height: 100%;
}

body {
  font-family: 'Roboto', sans-serif;
  margin: 0;
  overflow: hidden;
  padding: 0;
}

.BountyManager {
  color: white;
  height: 100%;
  left: 0;
  position: absolute;
  right: 0;
  text-align: center;
  top: 0;
}

.hex-background {
  background-color: #8500ff;
  background-image: url('/img/hexgrid_2x.png');
  background-size: 129px 149px;
}

.BountyManager-Intro {
  font-size: large;
}
