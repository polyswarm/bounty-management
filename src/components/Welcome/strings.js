const strings = {
  welcome: 'Welcome to Polyswarm,',
  decentralized: 'The worlds first decentralized marketplace',
  moreInfo: 'Simply upload files to create your own bounties and then let the microengines on Polyswarm analyze your file.',
  getStarted: 'Get Started!',
};

export default strings;
