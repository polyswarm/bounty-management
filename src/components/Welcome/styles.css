.Welcome {
  grid-column: 1 / 3;
  grid-row: 1 / 3;
  height: 100%;
}

.Welcome-Jumbo {
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: center;
  margin: 0 auto;
  max-width: 850px;
}

.Welcome-Jumbo > h1 {
  font-size: 5em;
  margin: 0;
}

.Welcome-Jumbo > h2 {
  font-size: 2em;
}

.Welcome-Jumbo > p {
  font-size: 16px;
  line-height: 16px;
  margin: 0 5px;
}

.Welcome-Jumbo > button {
  border-radius: 2px;
  margin: 20px 0;
  max-width: 180px;
  padding: 0;
}
