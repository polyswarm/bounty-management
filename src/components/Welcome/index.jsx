// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
// Bounty imports
import Button from '../Button';
// Component Imports
import strings from './strings';

class Welcome extends Component {
  constructor(props) {
    super(props);
    this.state = {
      accepted: false,
    };
    this.onClickHandler = this.onClickHandler.bind(this);
    this.onCheckedHandler = this.onCheckedHandler.bind(this);
  }

  render() {
    const {accepted} = this.state;
    return (
      <div className='Welcome'>
        <div className='Welcome-Jumbo'>
          <h1>{strings.welcome}</h1>
          <h2>{strings.moreInfo}</h2>
          <form>
            <input type="checkbox"
              id="accept"
              onChange={this.onCheckedHandler}
              value={accepted}/>
            <label htmlFor="accept">{strings.accept}</label>
          </form>
          <Button className='Welcome-Close'
            disabled={!accepted}
            cancel
            onClick={this.onClickHandler}>
            {strings.getStarted}
          </Button>
        </div>
      </div>
    );
  }

  onClickHandler() {
    const { props: { onClick } } = this;
    if (onClick) {
      onClick();
    }
  }

  onCheckedHandler(event) {
    const checked = event.target.checked;
    this.setState({accepted: checked});
  }
}

Welcome.propTypes = {
  onClick: PropTypes.func,
};
export default Welcome;
