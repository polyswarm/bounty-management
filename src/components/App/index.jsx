// Vendor imports
import React, { Component } from 'react';
// Bounty imports
import BountyCreate from '../BountyCreate';
import BountyInfo from '../BountyInfo';
import Sidebar from '../Sidebar';
import Header from '../Header';
import Welcome from '../Welcome';
// Component imports
import HttpApp from './http';
import config from './config';
import strings from './strings';
import './styles.css';

class App extends Component {
  constructor(props) {
    super(props);
    const {bounties, first} = this.preloadLocalStorage();
    this.http = new HttpApp(config.url, config.wsUrl);
    this.state = {
      isUnlocked: false,
      walletList: [],
      active: 0,
      bounties: bounties,
      create: false,
      first: first,
    };

    this.onAddBounty = this.onAddBounty.bind(this);
    this.onRemoveBounty = this.onRemoveBounty.bind(this);
    this.onSelectBounty = this.onSelectBounty.bind(this);
    this.onCreateBounty = this.onCreateBounty.bind(this);
    this.onCloseWelcome = this.onCloseWelcome.bind(this);
    this.onWalletChangeHandler = this.onWalletChangeHandler.bind(this);
    this.getData = this.getData.bind(this);
    this.getWallets = this.getWallets.bind(this);
  }

  componentDidUpdate(_, prevState) {
    const { state: { bounties } } = this;
    const { bounties: prevBounties } = prevState;
    const storageOutOfSync =  JSON.stringify(bounties) !== JSON.stringify(prevBounties);
    if(storageOutOfSync) {
      this.storeBounties(bounties);
    }
  }

  componentDidMount() {
    this.getWallets();
    this.getData();
  }

  render() {
    const {url} = config;
    const { state: { active, bounties, create, first, isUnlocked, walletList } } = this;

    return (
      <div className='App'>
        {first && (
          <Welcome onClick={this.onCloseWelcome}/>
        )}
        {!first && (
          <React.Fragment>
            <Sidebar bounties={bounties}
              active={active}
              remove={this.onRemoveBounty}
              select={this.onSelectBounty}/>
            <Header title={(bounties.length === 0 || create || active < 0) ? strings.create : bounties[active].guid}
              create={create}
              onClick={this.onCreateBounty}/>
            <div className='App-Content'>
              { (bounties.length === 0 || create || active < 0 ) && (
                <BountyCreate url={url}
                  isUnlocked={isUnlocked}
                  walletList={walletList}
                  onWalletChange={this.onWalletChangeHandler}
                  addBounty={this.onAddBounty}/>
              )}
              { !create && active >=0 && active < bounties.length && (
                <BountyInfo bounty={bounties[active]}/>
              )}
            </div>
          </React.Fragment>
        )}
      </div>
    );
  }

  onAddBounty(guid) {
    const bounties = this.state.bounties.slice();
    const bounty = {
      guid: guid,
      update: false,
      author: '',
      amount: '',
      artifactURI: '',
      expirationBlock: '',
      resolved: '',
      verdicts: '',
    };
    bounties.push(bounty);
    this.setState({bounties: bounties});
  }

  onCreateBounty() {
    this.setState({create: true, active: -1});
  }

  onCloseWelcome() {
    this.setState({first: false});
    this.markSeen();
  }

  onRemoveBounty(index) {
    const bounties = this.state.bounties.slice();
    if (index !== null && index >= 0 && index < bounties.length) {
      bounties.splice(index, 1);
      this.setState({bounties: bounties});
    }
  }

  onSelectBounty(index) {
    const bounties = this.state.bounties.slice();
    if (index !== null && index >= 0 && index < bounties.length) {
      const bounty = bounties[index];
      bounty.updated = false;
      bounties[index] = bounty;
      this.setState({active: index, create: false, bounties: bounties});
    }
  }

  onWalletChangeHandler(store) {
    this.setState({isUnlocked: store});
  }

  pullLocalBounties() {
    const bounties = this.state.bounties.slice();
    return bounties;
  };

  updateOnAssertion(assertion) {
    const bounties = this.state.bounties.slice();
    const guid = assertion.guid;
    const index = bounties.findIndex((bounty) => bounty.guid === guid);
    // let index = -1;
    // bounties.forEach((bounty, i) => {
    //   if (bounty.guid === guid) {
    //     index = i;
    //   }
    // });
    if (index >= 0) {
      const bounty = bounties[index];
      const a = {};
      a[assertion.author] = {
        bid: assertion.bid,
        metadata: assertion.metadata,
        verdicts: assertion.verdicts,
      };
      bounty.assertsions.push(a);
      const modified = bounty.artifacts.map((file, index) => {
        const f = file;
        if (!a.verdicts[index]) {
          f.good++;
        }
        f.total++;
        file.assertions.push({
          author: assertion.author,
          bid: assertion.bid,
          verdict: a.verdicts[index],
          metadata: assertion.metadata
        });
        return f;
      });
      bounty.artifacts = modified;
      bounty.update = true;
      bounties[index] = bounty;
    }
    this.setState({bounties: bounties});
  }

  getData() {
    const http = this.http;
    http.listenForAssertions(this.pullLocalBounties, this.updateOnAssertion);
    const bounties = this.state.bounties.slice();
    const promises = bounties.map((bounty) => {
      return http.getBounty(bounty)
        .then(b => {
          if (Object.keys(b) !== Object.keys(bounty)) {
            b.update = true;
          }
          return b;
        })
        .catch(() => bounty);
    });
    Promise.all(promises).then((values) => {
      // get updated state after download finishes
      const bounties = this.state.bounties.slice();
      values.forEach((value) => {
        const foundIndex = bounties.findIndex((bounty) => bounty.guid === value.guid);
        if (foundIndex >= 0) {
          bounties[foundIndex] = value;
        }
      });
      this.setState({bounties: bounties});
    });
  }

  getWallets() {
    const http = this.http;
    http.getWallets()
      .then(accounts => {
        this.setState({walletList: accounts});
      });

    http.getUnlockedWallet()
      .then((success) => this.setState({isUnlocked: success}));
  }

  storeBounties(bounties) {
    if (this.hasLocalStorage()) {
      localStorage.setItem('bounties', JSON.stringify(bounties));
    }
  }

  hasLocalStorage() {
    try {
      localStorage.setItem('x', 'y');
      localStorage.removeItem('x');
      return true;
    } catch(e) {
      return false;
    }
  }

  markSeen() {
    if (this.hasLocalStorage()) {
      localStorage.setItem('seen', JSON.stringify(true));
    }
  }

  preloadLocalStorage() {
    if (this.hasLocalStorage) {
      const bounties = JSON.parse(localStorage.getItem('bounties')) || [];
      const first = !JSON.parse(localStorage.getItem('seen'));
      return {bounties, first};
    } else {
      return {bounties: [], first: true};
    }
  }
}

export default App;
