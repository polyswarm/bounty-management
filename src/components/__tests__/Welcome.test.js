import React from 'react';
import {render, mount} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import Welcome from '../Welcome';
import App from '../App';

it('renders without crashing', () => {
  const wrapper = render(<Welcome />);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('should call onClick when button is clicked', () => {
  const onClick = jest.fn();
  const wrapper = mount(<Welcome onClick={onClick}/>);
  wrapper.setState({accepted: true});

  wrapper.find('button').simulate('click');

  expect(onClick).toHaveBeenCalledTimes(1);
});

it('should close when button is clicked', () => {
  const wrapper = mount(<App />);
  wrapper.setState({first: true});

  expect(wrapper.find('.Welcome')).toHaveLength(1);

  wrapper.find('.Welcome').find('input').simulate('change', { target: { checked: true } });
  wrapper.find('.Welcome').find('button').simulate('click');

  expect(wrapper.find('.Welcome')).toHaveLength(0);
});

it('should change accepted to true when checked', () => {
  const setState = jest.spyOn(Welcome.prototype, 'setState');
  const wrapper = mount(<Welcome/>);

  wrapper.find('input').simulate('change', { target: { checked: true } });

  expect(setState).toHaveBeenCalledWith({accepted: true});
});

it('should change accept to false when unchecked', () => {
  const setState = jest.spyOn(Welcome.prototype, 'setState');
  const wrapper = mount(<Welcome/>);
  wrapper.setState({accepted: true});
  setState.mockClear();

  wrapper.find('input').simulate('change', { target: { checked: false } });

  expect(setState).toHaveBeenCalledWith({accepted: false});
});

it('should enable the button when accepted is true', () => {
  const wrapper = mount(<Welcome/>);
  wrapper.setState({accepted: true});

  expect(wrapper.find('button').props().disabled).toBeFalsy();
});

it('should disable the button when accepted is false', () => {
  const wrapper = mount(<Welcome/>);
  wrapper.setState({accepted: false});

  expect(wrapper.find('button').props().disabled).toBeTruthy();
});
