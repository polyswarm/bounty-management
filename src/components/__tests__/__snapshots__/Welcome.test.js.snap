// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="Welcome"
>
  <div
    class="Welcome-Jumbo"
  >
    <h1>
      Welcome to Polyswarm,
    </h1>
    <h2>
      The first 
      <em>
        decentralized
      </em>
       antivirus marketplace.
    </h2>
    <p>
      Simply upload files to create your own bounties and then let the microengines on Polyswarm analyze your file.
    </p>
    <button
      class="Button"
    >
      Get Started!
    </button>
  </div>
</div>
`;
